File: Floating_Note_OpenGL/simple_vector_library.py
import numpy as np
import math
from random import *

class Py_Vector:
    def __init__(self, x = 0, y = 0, z = 0):
        self.x_component = x
        self.y_component = y
        self.z_component = z

    def copy(self):
        return(Py_Vector(self.x_component, self.y_component, self.z_component))

    def add(self, x, y=0, z=0):
        if(isinstance(x, Py_Vector)):
            self.x_component += x.x_component
            self.y_component += x.y_component
            self.z_component += x.z_component

        elif(isinstance(x, (np.ndarray, np.generic, list, tuple))):
            self.x_component += x[0]
            self.y_component += x[1]
            if(len(x) == 3):
                self.z_component += x[2]

        else:
            self.x_component += x
            self.y_component += y
            self.z_component += z

    def sub(self, x, y=0, z=0):
        if(isinstance(x, Py_Vector)):
            self.x_component -= x.x_component
            self.y_component -= x.y_component
            self.z_component -= x.z_component

        elif(isinstance(x, (np.ndarray, np.generic, list, tuple))):
            self.x_component -= x[0]
            self.y_component -= x[1]
            if(len(x) == 3):
                self.z_component -= x[2]

        else:
            self.x_component -= x
            self.y_component -= y
            self.z_component -= z

    def multi(self, scale):
        self.x_component *= scale
        self.y_component *= scale
        self.z_component *= scale

    def div(self, n):
        if(n == 0):
            print("Can't divide by zero")

        self.x_component /= n
        self.y_component /= n
        self.z_component /= n

    def mag(self):
        return math.sqrt(self.x_component * self.x_component
                        + self.y_component * self.y_component
                        + self.z_component * self.z_component)

    def dot(self, x, y=0, z=0):
        if(isinstance(x, Py_Vector)):
            return (self.x_component * x.x_component
                    + self.y_component * x.y_component
                    + self.z_component * x.z_component)

        return (self.x_component * x
                + self.y_component * y
                + self.z_component * z)


    def cross(self, vector):
        cross_x = (self.y_component * vector.z_component
                    - self.z_component * vector.y_component)
        cross_y = (self.z_component * vector.x_component
                    - self.x_component * vector.z_component)
        cross_z = (self.x_component * vector.y_component
                    - self.y_component * vector.x_component)

        return Py_Vector(cross_x, cross_y, cross_z)

    def dist(self, vector):
        return vector.copy.sub(self).mag()

    def normalize(self):
        if self.mag == 0:
            return self
        else:
            return self.div(self.mag())

    def limit(self, max):
        mSq = (self.x_component * self.x_component
                        + self.y_component * self.y_component
                        + self.z_component * self.z_component)
        if(mSq > max * max):
            self.div(math.sqrt(mSq))
            self.multi(max)

        return self

    def setMag(self, n):
        return self.normalize().multi(n)

    def heading(self):
        #Returns radians
        h = math.atan2(self.y_component, self.x_component)
        return h

    def rotate(self, a):
        newHeading = self.heading + a
        mag = self.mag()
        self.x_component = math.cos(newHeading) * setMag
        self.y_component = math.sin(newHeading) * setMag
        return self

    def angleBetween(self, vector):
        dotmagmag = self.dot(vector) / (self.mag() * vector.mag())
        angle = math.acos(min(1, max(-1, dotmagmag)))
        return angle

File: openFrameworks_testing/src/simple_vector_library.cpp
//
//  simple_vector_library.cpp
//  openFrameworks_testing
//
//  Created by Kevin Nottberg on 2/2/18.
//

#include "simple_vector_library.hpp"
#include <algorithm>
#include <math.h>

//2D Vector
SimpleVector::SimpleVector( float x, float y ) : x_comp(x), y_comp(y), z_comp(0) {};

//3D Vector
SimpleVector::SimpleVector( float x, float y, float z ) : x_comp(x), y_comp(y), z_comp(z) {};

//Addition Methods
void SimpleVector::add( SimpleVector &obj ) {
    this->x_comp += obj.x_comp;
    this->y_comp += obj.y_comp;
    this->z_comp += obj.z_comp;
}

void SimpleVector::add( float x, float y ) {
    this->x_comp += x;
    this->y_comp += y;
}

void SimpleVector::add( float x, float y, float z ) {
    this->x_comp += x;
    this->y_comp += y;
    this->z_comp += z;
}

//Subtraction Methods
void SimpleVector::sub( SimpleVector obj ) {
    this->x_comp -= obj.x_comp;
    this->y_comp -= obj.y_comp;
    this->z_comp -= obj.z_comp;
}

void SimpleVector::sub( float x, float y ) {
    this->x_comp -= x;
    this->y_comp -= y;
}

void SimpleVector::sub( float x, float y, float z ) {
    this->x_comp -= x;
    this->y_comp -= y;
    this->z_comp -= z;
}

//Multipication
void SimpleVector::multi( float scalar ) {
    this->x_comp *= scalar;
    this->y_comp *= scalar;
    this->z_comp *= scalar;
}

//Division
void SimpleVector::div( float scalar ) {
    if(scalar != 0.0) {
        this->x_comp /= scalar;
        this->y_comp /= scalar;
        this->z_comp /= scalar;
    }
}

//Dot-product methods
float SimpleVector::dot( SimpleVector obj ) {
    return (this->x_comp * obj.x_comp + this->y_comp * obj.y_comp, this->z_comp * obj.z_comp);
}

float SimpleVector::dot( float x, float y ) {
    return (this->x_comp * x + this->y_comp * y + this->z_comp * 0);
}

float SimpleVector::dot( float x, float y, float z ) {
    return (this->x_comp * x + this->y_comp * y + this->z_comp * z);
}

SimpleVector SimpleVector::cross( SimpleVector obj ) {
    float cross_x = (this->y_comp * obj.z_comp - this->z_comp * obj.y_comp);
    float cross_y = (this->z_comp * obj.x_comp - this->x_comp * obj.z_comp);
    float cross_z = (this->x_comp * obj.y_comp - this->y_comp * obj.z_comp);
    
    return SimpleVector(cross_x, cross_y, cross_z);
}

SimpleVector SimpleVector::normalize() {
    if(this->mag() == 0.0) {
        return *this;
    } else {
        SimpleVector new_vector = this->copy();
        new_vector.div(new_vector.mag());
        this->x_comp = new_vector.x_comp;
        this->y_comp = new_vector.y_comp;
    }
}

SimpleVector SimpleVector::rotate( float rotation ) {
    float new_heading = this->heading() + rotation;
    float mag = this->mag();
    
    this->x_comp = cos(new_heading) * mag;
    this->y_comp = sin(new_heading) * mag;
    
    return *this;
}

float SimpleVector::dist( SimpleVector obj ) {
    SimpleVector tmp = obj.copy();
    tmp.sub(*this);
    return tmp.mag();
}

void SimpleVector::limit( float lim ) {
    float mSq = (this->x_comp * this->x_comp +
                 this->y_comp * this->y_comp +
                 this->z_comp * this->z_comp);
    if(mSq > lim * lim) {
        this->div(sqrt(mSq));
        this->multi(lim);
    }
    
    return *this;
}

float SimpleVector::mag() {
    return sqrt(this->x_comp * this->x_comp +
                this->y_comp * this->y_comp +
                this->z_comp * this->z_comp);
}

void SimpleVector::setMag( float magn ) {
    this->normalize().multi(magn);
}

float SimpleVector::heading() {
    return atan2(this->y_comp, this->x_comp);
}

float SimpleVector::angleBetween( SimpleVector obj ) {
    float dot_mag_mag = this->dot(obj) / (this->mag() * obj.mag());
    return cos(fmin(1.0, fmax(-1.0, dot_mag_mag)));
}

SimpleVector SimpleVector::copy() {
    return SimpleVector(this->x_comp, this->y_comp, this->z_comp);
}













